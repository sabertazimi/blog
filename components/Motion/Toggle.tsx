import type { MotionProps } from '@components/utils'
import { AnimatePresence, cx, motion } from '@components/utils'
import type { KeyboardEvent, ReactNode } from 'react'
import { useCallback } from 'react'
import styles from './Toggle.module.css'

interface Props {
  isToggled: boolean
  onToggle?: () => void
  iconClose: ReactNode
  iconOpen: ReactNode
  shouldReduceMotion: boolean | null
  className?: string
  tabIndex?: number
}

const closeVariants: MotionProps['initial'] & MotionProps['exit'] = {
  rotate: 180,
  scale: 0,
  opacity: 0,
  transition: {
    type: 'tween',
    ease: 'easeInOut',
    duration: 0.3,
  },
}

const openVariants: MotionProps['animate'] = {
  rotate: 0,
  scale: 1,
  opacity: 1,
  transition: {
    type: 'tween',
    ease: 'easeInOut',
    duration: 0.3,
  },
}

function Toggle({
  isToggled,
  onToggle,
  iconClose,
  iconOpen,
  className,
  shouldReduceMotion,
  tabIndex = 0,
  ...props
}: Props): JSX.Element {
  const handleKeyDown = useCallback(
    (event: KeyboardEvent<HTMLDivElement>) => {
      if (event.key === 'Enter' && onToggle)
        onToggle()
    },
    [onToggle],
  )

  return (
    <div
      key="toggle-wrapper"
      data-testid="toggle-wrapper"
      tabIndex={tabIndex}
      onClick={onToggle}
      onKeyDown={handleKeyDown}
      className={cx(styles.wrapper, className)}
      {...props}
    >
      <AnimatePresence initial={false}>
        {!isToggled
          ? (
            <motion.span
              key="toggle-span-close"
              initial={closeVariants}
              animate={!shouldReduceMotion ? openVariants : closeVariants}
              exit={closeVariants}
              aria-current={!isToggled}
              className={styles.span}
            >
              {iconClose}
            </motion.span>
            )
          : (
            <motion.span
              key="toggle-span-open"
              initial={closeVariants}
              animate={!shouldReduceMotion ? openVariants : closeVariants}
              exit={closeVariants}
              aria-current={isToggled}
              className={styles.span}
            >
              {iconOpen}
            </motion.span>
            )}
      </AnimatePresence>
    </div>
  )
}

export default Toggle
