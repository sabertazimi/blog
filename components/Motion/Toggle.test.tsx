import { fireEvent, render, screen } from '@utils'
import Toggle from './Toggle'

describe('toggle', () => {
  const cases = [
    [false, false],
    [false, true],
    [true, false],
    [true, true],
  ]

  it.each(cases)(
    'should render animation correctly (snapshot)',
    (isToggled, shouldReduceMotion) => {
      const { container } = render(
        <Toggle
          isToggled={isToggled}
          onToggle={jest.fn()}
          iconClose={<div>Close</div>}
          iconOpen={<div>Open</div>}
          shouldReduceMotion={shouldReduceMotion}
        />,
      )

      expect(container).toMatchSnapshot()
    },
  )

  it('should invoke `onToggle` when clicked', () => {
    const mockToggle = jest.fn()

    render(
      <Toggle
        isToggled={false}
        onToggle={mockToggle}
        iconClose={<div>Close</div>}
        iconOpen={<div>Open</div>}
        shouldReduceMotion={false}
      />,
    )

    fireEvent.click(screen.getByTestId('toggle-wrapper'))

    expect(mockToggle).toHaveBeenCalledTimes(1)
  })

  it('should invoke `onToggle` when `Enter` key down', () => {
    const mockToggle = jest.fn()

    render(
      <Toggle
        isToggled={false}
        onToggle={mockToggle}
        iconClose={<div>Close</div>}
        iconOpen={<div>Open</div>}
        shouldReduceMotion={false}
      />,
    )

    fireEvent.keyDown(screen.getByTestId('toggle-wrapper'), { key: 'Enter' })

    expect(mockToggle).toHaveBeenCalledTimes(1)
  })

  it('should not invoke `onToggle` when `Tab` key down', () => {
    const mockToggle = jest.fn()

    render(
      <Toggle
        isToggled={false}
        onToggle={mockToggle}
        iconClose={<div>Close</div>}
        iconOpen={<div>Open</div>}
        shouldReduceMotion={false}
      />,
    )

    fireEvent.keyDown(screen.getByTestId('toggle-wrapper'), { key: 'Tab' })

    expect(mockToggle).not.toBeCalled()
  })
})
