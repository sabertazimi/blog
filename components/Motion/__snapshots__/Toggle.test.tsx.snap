// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should render animation correctly (snapshot) 1`] = `
<div>
  <div
    class="wrapper"
    data-testid="toggle-wrapper"
    tabindex="0"
  >
    <span
      aria-current="true"
      class="span"
      style="opacity: 1; transform: none;"
    >
      <div>
        Close
      </div>
    </span>
  </div>
</div>
`;

exports[`Toggle should render animation correctly (snapshot) 2`] = `
<div>
  <div
    class="wrapper"
    data-testid="toggle-wrapper"
    tabindex="0"
  >
    <span
      aria-current="true"
      class="span"
      style="opacity: 0; transform: scale(0) rotate(180deg);"
    >
      <div>
        Close
      </div>
    </span>
  </div>
</div>
`;

exports[`Toggle should render animation correctly (snapshot) 3`] = `
<div>
  <div
    class="wrapper"
    data-testid="toggle-wrapper"
    tabindex="0"
  >
    <span
      aria-current="true"
      class="span"
      style="opacity: 1; transform: none;"
    >
      <div>
        Open
      </div>
    </span>
  </div>
</div>
`;

exports[`Toggle should render animation correctly (snapshot) 4`] = `
<div>
  <div
    class="wrapper"
    data-testid="toggle-wrapper"
    tabindex="0"
  >
    <span
      aria-current="true"
      class="span"
      style="opacity: 0; transform: scale(0) rotate(180deg);"
    >
      <div>
        Open
      </div>
    </span>
  </div>
</div>
`;
